lsp_test.go

```diff
@@ -47,6 +49,361 @@ func TestDocumentOpen(t *testing.T) {
+								},
+							},
+						},
+						{
+							Message:  "terramate schema error",
+							Severity: lsp.DiagnosticSeverityError,
+							Range: lsp.Range{
+								Start: lsp.Position{
+									Line:      4,
+									Character: 2,
+								},
+								End: lsp.Position{
+									Line:      4,
+									Character: 3,
+								},
+							},
+						},
+						{
+							Message:  "terramate schema error",
+							Severity: lsp.DiagnosticSeverityError,
+							Range: lsp.Range{
+								Start: lsp.Position{
+									Line:      6,
+									Character: 1,
+								},
+								End: lsp.Position{
+									Line:      6,
+									Character: 10,
+								},
+							},
+						},
+						{
+							Message:  "terramate schema error",
+							Severity: lsp.DiagnosticSeverityError,
+							Range: lsp.Range{
+								Start: lsp.Position{
+									Line:      11,
+									Character: 1,
+								},
+								End: lsp.Position{
+									Line:      11,
+									Character: 2,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple errors in the workspace",
+			change: change{
+				file: "terramate.tm",
+				text: "",
+			},
+			layout: []string{
+				`f:bug1.tm:terramate {
+					a = 1
+					b = 2
+				}	
+					`,
+			},
+			want: []WantDiagParams{
+				{
+					URI: "bug1.tm",
+					Diagnostics: []WantDiag{
+						{
+							Message:  "terramate schema error",
+							Severity: lsp.DiagnosticSeverityError,
+							Range: lsp.Range{
+								Start: lsp.Position{
+									Line:      1,
+									Character: 5,
+								},
+								End: lsp.Position{
+									Line:      1,
+									Character: 6,
+								},
+							},
+						},
+						{
+							Message:  "terramate schema error",
+							Severity: lsp.DiagnosticSeverityError,
+							Range: lsp.Range{
+								Start: lsp.Position{
+									Line:      2,
+									Character: 5,
+								},
+								End: lsp.Position{
+									Line:      2,
+									Character: 6,
+								},
+							},
+						},
+					},
+				},
+				{
+					URI:         "terramate.tm",
+					Diagnostics: []WantDiag{},
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			f := test.Setup(t, tc.layout...)
+			f.Editor.CheckInitialize()
+
+			f.Editor.Change(tc.change.file, tc.change.text)
+			for i := 0; i < len(tc.want); i++ {
+				want := tc.want[i]
+
+				// fix the wanted path as it depends on the sandbox root.
+				want.URI = uri.File(filepath.Join(f.Sandbox.RootDir(), string(want.URI)))
+				select {
+				case gotReq := <-f.Editor.Requests:
+					assert.EqualStrings(t, lsp.MethodTextDocumentPublishDiagnostics,
+						gotReq.Method())
+
+					var gotParams lsp.PublishDiagnosticsParams
+					assert.NoError(t, json.Unmarshal(gotReq.Params(), &gotParams))
+
+					assert.EqualInts(t,
+						len(gotParams.Diagnostics), len(want.Diagnostics),
+						"number of diagnostics mismatch")
+
+					assert.Partial(t, gotParams, want, "diagnostic mismatch")
+				case <-time.After(10 * time.Millisecond):
```

Comment:
hmm to avoid possible failures because something got unexpectedly slow, couldn't we have a bigger timeout ? Usually on tests I never do things like that because go tests have a timeout intrinsic to them anyway (although it is usually much bigger than this). We would be able to fail faster if something is wrong, but also seems like extra work + easier to fail when we shouldn't (just got slower, host has a lot of load on it, etc). :thinking:

Reply:
All the tests ran in milliseconds but if trying to read more requests than ever available then it blocks and we have to wait for the go timeout (10s at the moment). The current test infrastructure makes it easy to block forever because we read as much requests as is in the wanted list, so very often I misinterpret the testcase and tests block (and even CTRL-C doesn't work as we didnt setup signals) so it's annoying... As I'm dumb, this happens a lot haha

So yeah, I can revert this but I'll probably do it again (just for helping myself) next time I work with those tests :smile: 

Other than that... I do believe we should try to avoid reading from unbuffered channel without timeouts in real-world software, so for me it looks like we should do the same in tests. We know go-test has timeout, but the tests can be used as example of real world code, so blocking forever is not nice. wdyt?

I don't like extra work, extra lines of code without real need, but this is very annoying for me =(

I see your point about go-test already have a timeout, so adding an arbitrary timeout here, depending on the value, could never fire because the go timeout has priority, but I think this is ok. The same thing you said about the loading of the machine also apply to the go test timeout, so someone would also need to change the code (the Makefile `test` rule) if the machine has a big load.

Reply:
> Another thing that I don't get is why unbuffered channels should be treated differently...
sorry I meant "unbuffered channels" == "channel without a buffered data".

I know it simplifies the test, but blocking forever as a default seems off to me.
I will remove this for now as we need to talk more about it.

Reply:
My argument is that in a test you are never blocking forever. It is different from doing this even on a small Go program and then it hangs forever, go test always exits, it is enforced by the tool, so you are only making it fail faster than the timeout provided on go test + maybe the error message could be more clear (or not, depends). But I don't see it as "Im writing a program that will hang forever".

I didn't get "sorry I meant "unbuffered channels" == "channel without a buffered data"." AFAIK there is no safe way to be sure a channel has data buffered or not (without races etc), the only way would be using a select like you did (with a timeout ou a default), so you need to take the same amount of care independent if a channel is buffered or not :thinking:

Reply:
I finally got the "unbuffered channel" thing and that it is not related to actual unbuffered/buffered channels, but a channel that is ready for a read operation :laughing: . Sorry for the noise.